### PR TITLE
fix state extraction from batch when fault-tolerant training

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes a bug with extracting the state from a data-queue batch when fault-tolerant training is enabled. 
In #8950 we moved the state extraction directly into the patched [iterator-next function](https://github.com/PyTorchLightning/pytorch-lightning/blob/c86e6cf1c4544f87b98b10813d0462875bd01cd9/pytorch_lightning/utilities/auto_restart.py#L566), but left behind some code that would still try to extract the state later on in an iterator at a higher level. 
This lead to a silent conversion of a batch to a list, but the bug would only occur when the **batch is a dict structure of data**.
The dead code resulting from this fix gets removed in this PR.

TODO: 
- [x] update test
- [x] confirm with test user

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
   **NOT NECESSARY: it was not included in any release.**

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/auto_restart.py [hunk 1]
```diff
@@ -354,72 +354,6 @@ def __iter__(self) -> Iterator:
...
-        else:
-            for attr_cache, state_dict in zip(iterator_state_dict, iterator._sampler_state_dict_cache):
-                for k, v in state_dict.items():
-                    attr_cache[k].update(v)
-            iterator._sampler_state_dict_cache = sampler_state_dict
-
-    @staticmethod
-    def _sanitize_batch_from_sampler_state(data: Any, state_dicts: List):
-        """
-        This function is used to remove the sampler state dict from provided data batch.
```
Comment (awaelchli) [inline pytorch_lightning/utilities/auto_restart.py:380]: TODO: move this documentation, we probably want to keep it

**Review by tchaton — APPROVED** [review]
Great catch !

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
